### PR TITLE
Analytics enable reset visible event

### PR DESCRIPTION
**Pull request by zhouyx**
For #11987 
update doc in separate PR.

### CODE REVIEW

File: extensions/amp-analytics/0.1/visibility-model.js [hunk 1]
```diff
@@ -115,6 +154,68 @@ export class VisibilityModel {
...
+  disposeOrReset() {
+    // We may need a maxIntervalWindow to stop visible event even with reset true
+    if (!this.reset_) {
+      this.dispose();
+      return;
+    }
+    if (this.resetInterval_) {
+      const now = Date.now();
+      const interval = now - this.firstVisibleTime_;
+      const diff = Math.max(0, this.resetInterval_ - interval);
```
Comment (jonkeller) [inline extensions/amp-analytics/0.1/visibility-model.js]: Maybe rename to something like:
timeUntilResetInterval
timeUntilResetDue
millisecondsUntilReset
etc?

File: extensions/amp-analytics/0.1/visibility-model.js [hunk 2]
```diff
@@ -115,6 +154,68 @@ export class VisibilityModel {
...
+      this.ready_ = false;
+      dev().assert(!this.scheduleResetId_, 'Should not reset twice');
+      this.scheduleResetId_ = setTimeout(() => {
+        this.refresh_();
+        this.setReady(true);
+      }, diff);
+    } else {
+      // Reset after element falls out of [minPercentage, maxPercentage]
```
Comment (jonkeller) [inline extensions/amp-analytics/0.1/visibility-model.js]: Nit: Is (minPercentage, maxPercentage] more correct?

File: test/manual/amp-analytics-visible-reset.amp.html [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+              "selector": "#ad2",
+              "visiblePercentageMin": "0",
+              "reset": true
+            }
+          }
+        }
+      }
+    </script>
+  </amp-analytics>
+  <h2>fake3pAd send render-start w/ width & height</h2>
```
Comment (jonkeller) [inline test/manual/amp-analytics-visible-reset.amp.html]: Nit: Replacing ampersand char with ampersand HTML entity would be more correct, but it's a test so not too concerned :)

File: test/manual/amp-analytics-visible-reset.amp.html [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        type="_ping_"
+        data-url='https://lh3.googleusercontent.com/pSECrJ82R7-AqeBCOEPGPM9iG9OEIQ_QXcbubWIOdkY=w400-h300-no-n'
+        data-valid='true'
+        data-ad-width=300
+        data-ad-height=250>
+    </amp-ad>
+
+
+  </div>
+</body>
```
Comment (jonkeller) [inline test/manual/amp-analytics-visible-reset.amp.html]: &lt;/html&gt;

File: extensions/amp-analytics/0.1/test/test-visibility-model.js [hunk 5]
```diff
@@ -144,8 +152,32 @@ describes.sandboxed('VisibilityModel', {}, () => {
...
+      // Accept boolean
+      expect(getReset({reset: true}).reset).to.be.true;
+      expect(getReset({reset: true}).resetInterval).to.be.null;
+      expect(getReset({reset: 'true'}).reset).to.be.false;
+      expect(getReset({reset: 'invalid'}).reset).to.be.false;
+
+      // Accept number
+      expect(getReset({reset: '200'}).reset).to.be.true;
+      expect(getReset({reset: 200}).reset).to.be.true;
+      expect(getReset({reset: 199}).reset).to.be.false;
```
[the remark this answers is not in the document]
Comment (zhouyx) [reply extensions/amp-analytics/0.1/test/test-visibility-model.js]: done

File: extensions/amp-analytics/0.1/visibility-model.js [hunk 6]
```diff
@@ -115,6 +154,68 @@ export class VisibilityModel {
...
+    this.waitToRefresh_ = false;
+
+    /** @private {?number} */
+    this.scheduleResetId_ = null;
+  }
+
+  /**
+   * Refresh counter on visible reset.
+   * TODO: Right now all state value are scoped state values that gets reset.
+   * We may need to add support to global state values, that never reset like globalTotalVisibleTime.
```
[the remark this answers is not in the document]
Comment (zhouyx) [reply extensions/amp-analytics/0.1/visibility-model.js]: fixed

File: extensions/amp-analytics/0.1/visibility-model.js [hunk 7]
```diff
@@ -115,6 +154,68 @@ export class VisibilityModel {
...
+   * Note: loadTimeVisibility is an exception.
+   * @private
+   */
+  refresh_() {
+    this.eventPromise_ = new Promise(resolve => {
```
[the remark this answers is not in the document]
Comment (zhouyx) [reply extensions/amp-analytics/0.1/visibility-model.js]: There shouldn't be. But I added check to `this.eventResolver_ == null` to make sure that `refresh_` is only called after the previous `eventPromise` is resolved.

File: extensions/amp-analytics/0.1/visibility-model.js (hunk 1 above)
Comment (zhouyx) [reply extensions/amp-analytics/0.1/visibility-model.js]: changed name.

File: extensions/amp-analytics/0.1/visibility-model.js (hunk 2 above)
Comment (zhouyx) [reply extensions/amp-analytics/0.1/visibility-model.js]: you're right! fixed

File: test/manual/amp-analytics-visible-reset.amp.html (hunk 3 above)
Comment (zhouyx) [reply test/manual/amp-analytics-visible-reset.amp.html]: removed : )

File: test/manual/amp-analytics-visible-reset.amp.html (hunk 4 above)
Comment (zhouyx) [reply test/manual/amp-analytics-visible-reset.amp.html]: fixed

**Comment by jonkeller** [comment]
Changes in response to my review comments all LGTM - thanks!

**Pull request merged**